File: haskell/app/AOC2024/Day02.hs
module AOC2024.Day02
  ( part1,
    part2,
  )
where

import Data.Text qualified as T

part1 :: T.Text -> Int
part1 input = length . filter isReportSafe $ parseReports input

part2 :: T.Text -> Int
part2 input = length . filter isAnyReportVarientSafe $ parseReports input
  where
    isAnyReportVarientSafe xs = any isReportSafe [take i xs ++ drop (i + 1) xs | i <- [0 .. length xs - 1]]

isReportSafe :: [Int] -> Bool
isReportSafe arr = isMonotonic arr && adjLvlDiff arr
  where
    isMonotonic (x : y : xs) = all ((== compare x y) . uncurry compare) (zip (y : xs) xs)
    isMonotonic [_] = True
    isMonotonic [] = True

    adjLvlDiff xs@(_ : rest) = all withinRange (zip xs rest)
    adjLvlDiff [] = True

    withinRange (a, b) = let d = abs (a - b) in d >= 1 && d <= 3

parseReports :: T.Text -> [[Int]]
parseReports input = [[read $ T.unpack x | x <- T.split (== ' ') reportStr] | reportStr <- T.lines input]

File: python/aoc2024/Day02.py
def adjLvlDiff(report):
  def withinRange(a, b):
    return 1 <= abs(a - b) <= 3
  return all([withinRange(report[i], report[i + 1]) for i in range(len(report) - 1)])


def isMonotonic(report):
  def getDir(a, b):
    return 'inc' if a < b else 'dec'

  def sameDir(a, b):
    return getDir(a, b) == initDir

  initDir = getDir(report[0], report[1])
  return all([sameDir(report[i], report[i + 1]) for i in range(1, len(report) - 1)])


def isReportSafe(report):
  return isMonotonic(report) and adjLvlDiff(report)


def isReportVarientSafe(report):
  return any([isReportSafe(report[:i] + report[i + 1:]) for i in range(len(report))])


# Main prog
with open("input.txt", 'r') as infile:
  data = infile.read().splitlines()

reports = [[int(lvl) for lvl in reportStr.split()] for reportStr in data]

safeReports = [r for r in reports if isReportSafe(r)]
print("part1: " + str(len(safeReports)))

toleratedReports = [r for r in reports if isReportVarientSafe(r)]
print("part2: " + str(len(toleratedReports)))
